src/semant/findescape.sml

```diff
@@ -0,0 +1,82 @@
+structure FindEscape: sig val findEscape: Absyn.exp -> unit
+                      end =
+struct
+  type depth = int
+  type escEnv = (depth * bool ref) Symbol.table
+
+  fun traverseVar(env:escEnv, d:depth, s:Absyn.var) : unit =
+    (case s of
+        Absyn.SimpleVar(sym,pos) =>
+          (case Symbol.look(env,sym) of
+              SOME(d',esc) => if d > d' then esc := true else ()
+            | NONE => ())
+      | Absyn.FieldVar(var,sym,pos) => (traverseVar(env,d,var))
+      | Absyn.SubscriptVar(var,exp,pos) =>
+        (traverseExp(env,d,exp);
+         traverseVar(env,d,var)))
+
+  and traverseExp(env:escEnv, d:depth, s:Absyn.exp) : unit =
+    (case s of
+        Absyn.VarExp(v) => traverseVar(env,d,v)
+      | Absyn.NilExp => ()
+      | Absyn.IntExp(_) => ()
+      | Absyn.StringExp(_) => ()
+      | Absyn.CallExp({args,func,pos}) =>
+        foldl (fn (exp,acc) => traverseExp(env,d,exp)) () args
+      | Absyn.OpExp({left,right,oper,pos}) =>
+        foldl (fn (exp,acc) => traverseExp(env,d,exp)) () [left,right]
+      | Absyn.RecordExp({fields,typ,pos}) =>
+        foldl (fn ((symbol,exp,pos),acc) => traverseExp(env,d,exp)) () fields
+      | Absyn.SeqExp(seq) =>
+        foldl (fn ((exp,pos),acc) => traverseExp(env,d,exp)) () seq
+      | Absyn.AssignExp({var,exp,pos}) =>
+        (traverseVar(env,d,var);
+         traverseExp(env,d,exp))
+      | Absyn.IfExp({test,then',else',pos}) =>
+        (traverseExp(env,d,test);
+         traverseExp(env,d,then');
+         (case else' of
+              SOME(exp) => traverseExp(env,d,exp)
+            | NONE => ()))
+      | Absyn.WhileExp({test,body,pos}) =>
+        (traverseExp(env,d,test);
+         traverseExp(env,d,body))
+      | Absyn.ForExp({var,escape,lo,hi,body,pos}) =>
+        (escape := false; (* TODO: I think this should be false, but if so, it should be changed in the parser, not here. This applies to all of the escapes *)
```

Comment:
@soelgary can you elaborate on why this should be false, although I think it makes sense to me. Is the idea that the loop variable never escapes because it's defined in place?